### PR TITLE
dm: support dm worker incremental validator cutover switch

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflow/issues/9868

### What is changed and how it works?
Part of adding interfaces to support worker incremental switch

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: dm/syncer/data_validator.go [hunk 1]
```diff
@@ -954,7 +955,11 @@ func (v *DataValidator) processRowsEvent(header *replication.EventHeader, ev *re
 
 func (v *DataValidator) checkAndPersistCheckpointAndData(loc binlog.Location) error {
 	metaFlushInterval := v.cfg.ValidatorCfg.MetaFlushInterval.Duration
-	if time.Since(v.lastFlushTime) > metaFlushInterval {
+	needCutOver := v.cutOverLocation != nil && binlog.CompareLocation(*v.cutOverLocation, loc, v.cfg.EnableGTID) <= 0
```
Comment (okJiang) [inline dm/syncer/data_validator.go:958]: if `v.cutOverLocation` is nil, validator will never `persistCheckpointAndData`?

File: dm/master/server.go [hunk 2]
```diff
@@ -3218,3 +3218,72 @@ func genValidationWorkerErrorResp(req *workerrpc.Request, err error, logMsg, wor
...
+			}, nil
+		}
+	}
+
+	workerReq := workerrpc.Request{
+		Type: workerrpc.CmdUpdateValidation,
+		UpdateValidation: &pb.UpdateValidationWorkerRequest{
+			TaskName:   req.TaskName,
+			BinlogPos:  req.BinlogPos,
+			BinlogGTID: req.BinlogGTID,
```
Comment (okJiang) [inline dm/master/server.go:3247]: How about adding a check for args that `BinlogPos` is empty and `EnableGTID` is false or `BinlogGTID` is empty and `EnableGTID` is true?

**Review by okJiang — COMMENTED** [review]
rest lgtm

File: dm/syncer/data_validator.go (hunk 1 above)
Comment (lichunzhu) [reply dm/syncer/data_validator.go:958]: sorry I make a mistake. The following condition should be `or`.

File: dm/master/server.go (hunk 2 above)
Comment (lichunzhu) [reply dm/master/server.go:3247]: I will fix it in next PR

File: dm/syncer/data_validator.go [hunk 3]
```diff
@@ -192,11 +194,12 @@ type DataValidator struct {
...
 	persistHelper      *validatorPersistHelper
 
-	validateInterval time.Duration
-	checkInterval    time.Duration
-	cutOverLocation  *binlog.Location
-	workers          []*validateWorker
-	workerCnt        int
+	validateInterval      time.Duration
+	checkInterval         time.Duration
+	cutOverLocationAtomic atomic.Pointer[binlog.Location]
```
Comment (okJiang) [inline dm/syncer/data_validator.go:199]:
```suggestion
	cutOverLocation atomic.Pointer[binlog.Location]
```

prefer this, `Atomic` is excessive.

File: dm/syncer/data_validator.go (hunk 1 above)
Comment (okJiang) [reply dm/syncer/data_validator.go]: Is this fixed?

**Review by okJiang — COMMENTED** [review]
rest lgtm

File: dm/syncer/data_validator.go (hunk 1 above)
Comment (lichunzhu) [reply dm/syncer/data_validator.go]: Yes. `&&` is revised to `||` now

**Review by okJiang — APPROVED** [review]

File: dm/ctl/master/update_validation.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	}
+	// we use taskName="" to represent we do operation on all task, so args.allTask don't need to pass
+	req := &pb.UpdateValidationRequest{
+		TaskName:   args.taskName,
+		Sources:    args.sources,
+		BinlogPos:  args.cutoverBinlogPos,
+		BinlogGTID: args.cutoverBinlogGTID,
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (lance6716) [inline dm/ctl/master/update_validation.go:60]: seems can directly use context.Background()

File: dm/ctl/master/update_validation.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func NewUpdateValidationCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "update [task-name]",
+		Short: "update validation config of the completeness of the data",
+		RunE:  updateValidation,
+	}
+	cmd.Flags().String("cutover-binlog-pos", "", "specify the cutover binlog name for validation, should include binlog name and pos in brackets, e.g. '(mysql-bin.000001, 5989)'")
+	cmd.Flags().String("cutover-binlog-gtid", "", "specify the cutover binlog gtid for validation, only valid when source config's gtid is enabled, e.g. '1642618e-cf65-11ec-9e3d-0242ac110002:1-30'")
```
Comment (lance6716) [inline dm/ctl/master/update_validation.go:43]: >cutover-binlog-gtid

in fact it's gtid set (gtid is one transaction like 1642618e-cf65-11ec-9e3d-0242ac110002:30). But seems in other places we also call it XXXgtid

File: dm/proto/dmworker.proto [hunk 6]
```diff
@@ -500,3 +502,10 @@ enum ValidationErrOp {
   ResolveErrOp = 2;
   ClearErrOp = 3;
 }
+
+
+message UpdateValidationWorkerRequest {
+    string taskName = 1;
+    string binlogPos = 2; // binlog-pos (that's file:pos format)
```
Comment (lance6716) [inline dm/proto/dmworker.proto:509]:
```suggestion
    string binlogPos = 2; // binlog-pos (that's (file:pos) format)
```

File: dm/syncer/data_validator.go [hunk 7]
```diff
@@ -1309,6 +1318,29 @@ func (v *DataValidator) OperateValidatorError(validateOp pb.ValidationErrOp, err
...
 }
 
+func (v *DataValidator) UpdateValidator(req *pb.UpdateValidationWorkerRequest) error {
+	var (
+		pos *mysql.Position
+		gs  mysql.GTIDSet
+		err error
+	)
+	if len(req.BinlogPos) > 0 {
+		pos, err = binlog.VerifyBinlogPos(req.BinlogGTID)
```
Comment (lance6716) [inline dm/syncer/data_validator.go:1328]:
```suggestion
		pos, err = binlog.VerifyBinlogPos(req.BinlogPos)
```

better to add basic test about this feature

File: dm/ctl/master/update_validation.go (hunk 4 above)
Comment (lichunzhu) [reply dm/ctl/master/update_validation.go]: addressed in https://github.com/pingcap/tiflow/pull/9867/commits/eaf71b5565ca33e6483782a99d4fa462cf1bbd6b

File: dm/ctl/master/update_validation.go (hunk 5 above)
Comment (lichunzhu) [reply dm/ctl/master/update_validation.go:43]: We also use `binlog-gtid` in `TaskConfig.Meta` configuration. I think we just keep the same now?

Comment (lance6716) [reply dm/ctl/master/update_validation.go:43]: lgtm

File: dm/ctl/master/update_validation.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+	return cmd
+}
+
+func updateValidation(cmd *cobra.Command, _ []string) error {
+	args, msg, ok := parseValidationUpdateArgs(cmd)
+	if !ok {
+		return printUsageAndFailWithMessage(cmd, msg)
+	}
+	// we use taskName="" to represent we do operation on all task, so args.allTask don't need to pass
```
Comment (D3Hunter) [inline dm/ctl/master/update_validation.go:52]: we have a err `"`task-name` should be set"`, what this comment mean, pass task-name as `""`?

File: dm/master/workerrpc/rawgrpc.go [hunk 9]
```diff
@@ -124,6 +124,8 @@ func callRPC(ctx context.Context, client pb.WorkerClient, req *Request) (*Respon
 		resp.GetValidationError, err = client.GetValidatorError(ctx, req.GetValidationError)
 	case CmdOperateValidationError:
 		resp.OperateValidationError, err = client.OperateValidatorError(ctx, req.OperateValidationError)
+	case CmdUpdateValidation:
+		resp.UpdateValidation, err = client.UpdateValidator(ctx, req.UpdateValidation)
```
Comment (D3Hunter) [inline dm/master/workerrpc/rawgrpc.go:128]: the feature is related to cutover, but method named as `UpdateValidator`, sounds like update validator config online

File: dm/syncer/data_validator.go [hunk 10]
```diff
@@ -954,7 +958,12 @@ func (v *DataValidator) processRowsEvent(header *replication.EventHeader, ev *re
 
 func (v *DataValidator) checkAndPersistCheckpointAndData(loc binlog.Location) error {
 	metaFlushInterval := v.cfg.ValidatorCfg.MetaFlushInterval.Duration
-	if time.Since(v.lastFlushTime) > metaFlushInterval {
+	cutOverLocation := v.cutOverLocation.Load()
+	needCutOver := cutOverLocation != nil && binlog.CompareLocation(*cutOverLocation, loc, v.cfg.EnableGTID) <= 0
```
Comment (D3Hunter) [inline dm/syncer/data_validator.go:962]: validator never validate before syncer's flushedPos, see waitSyncerSynced, maybe we should trigger syncer flush too in cutover

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: dm/ctl/master/update_validation.go (hunk 8 above)
Comment (lichunzhu) [reply dm/ctl/master/update_validation.go:52]: the argguments are just copied from other functions. already removed

File: dm/master/workerrpc/rawgrpc.go (hunk 9 above)
Comment (lichunzhu) [reply dm/master/workerrpc/rawgrpc.go:128]: It's copied from spec. Any suggestion?

File: dm/syncer/data_validator.go (hunk 10 above)
Comment (lichunzhu) [reply dm/syncer/data_validator.go:962]: Get your point. What about in next PR? It's hard to catch this now.

Comment (D3Hunter) [reply dm/syncer/data_validator.go:962]: lgtm

**Pull request merged**